theano/tensor/opt.py

```diff
@@ -1269,6 +1269,59 @@ def apply(self, fgraph):
                                    0.1, 'fast_run', 'fast_compile')
 
 
+@gof.local_optimizer([T.Elemwise])
+def local_fill_sink(node):
+    """
+    f(fill(a, b), fill(c, d), e) -> fill(a, fill(c, f(b, d, e)))
+
+    f need to be an elemwise
+    """
+    if (not hasattr(node, 'op') or
+            not isinstance(node.op, T.Elemwise) or
+            node.op == T.fill):
+        return False
+    models = []
+    inputs = []
+    for input in node.inputs:
+        if input.owner and input.owner.op == T.fill:
+            models.append(input.owner.inputs[0])
+            inputs.append(input.owner.inputs[1])
+        else:
+            inputs.append(input)
+    if not models:
+        return False
+    c = node.op(*inputs)
+    for model in models:
+        c = T.fill(model, c)
```

Comment:
I think I understand what is missing. This line create more fill node. The opt local_fill_to_alloc also remove useless fill, not just the opt local_useless_fill. Instead of doing this ping/pong between this opt and local_fill_to_alloc, can you try to introduce here model only when it is not a useless fill?

replace the loop with this:
```
    for model in models:
        if model.type != c.type:
            c = T.fill(model, c)
```

Can you add this, and rebase this PR on top of 3232 to don't have them collide and time again?